src/devices/sound/gb.cpp

```diff
@@ -1323,50 +1317,49 @@ void gameboy_sound_device::sound_stream_update(sound_stream &stream, std::vector
 		s32 left = 0;
 		s32 right = 0;
 
-		/* Mode 1 - Wave with Envelope and Sweep */
-		if (m_snd_1.on)
+		/* Channel 1 - Wave with Envelope and Sweep */
+		if (m_snd[0].on)
 		{
-			sample = m_snd_1.signal * m_snd_1.envelope_value;
-
-			if (m_snd_control.mode1_left)
+			sample = convert_output(m_snd[0].signal * m_snd[0].envelope_value);
+			if (m_snd_control.chan_left[0])
 				left += sample;
-			if (m_snd_control.mode1_right)
+			if (m_snd_control.chan_right[0])
 				right += sample;
 		}
 
-		/* Mode 2 - Wave with Envelope */
-		if (m_snd_2.on)
+		/* Channel 2 - Wave with Envelope */
+		if (m_snd[1].on)
 		{
-			sample = m_snd_2.signal * m_snd_2.envelope_value;
-			if (m_snd_control.mode2_left)
+			sample = convert_output(m_snd[1].signal * m_snd[1].envelope_value);
+			if (m_snd_control.chan_left[1])
 				left += sample;
-			if (m_snd_control.mode2_right)
+			if (m_snd_control.chan_right[1])
 				right += sample;
 		}
 
-		/* Mode 3 - Wave patterns from WaveRAM */
-		if (m_snd_3.on)
+		/* Channel 3 - Wave patterns from WaveRAM */
+		if (m_snd[2].on)
 		{
-			sample = m_snd_3.signal;
-			if (m_snd_control.mode3_left)
+			sample = convert_output(m_snd[2].signal);
+			if (m_snd_control.chan_left[2])
 				left += sample;
-			if (m_snd_control.mode3_right)
+			if (m_snd_control.chan_right[2])
 				right += sample;
 		}
 
-		/* Mode 4 - Noise with Envelope */
-		if (m_snd_4.on)
+		/* Channel 4 - Noise with Envelope */
+		if (m_snd[3].on)
 		{
-			sample = m_snd_4.signal * m_snd_4.envelope_value;
-			if (m_snd_control.mode4_left)
+			sample = convert_output(m_snd[3].signal * m_snd[3].envelope_value);
+			if (m_snd_control.chan_left[3])
 				left += sample;
-			if (m_snd_control.mode4_right)
+			if (m_snd_control.chan_right[3])
 				right += sample;
 		}
 
 		/* Adjust for master volume */
-		left *= m_snd_control.vol_left;
-		right *= m_snd_control.vol_right;
+		left *= 1 + m_snd_control.vol_left;
+		right *= 1 + m_snd_control.vol_right;
 
 		/* Update the buffers */
 		outputl.put_int(sampindex, left, 32768 / 64);
```

Comment:
I’m not sure this will be giving the correct scaled values now.

* The range of the native channel outputs is 0x0..0x0f, corresponding to normalised 1..-1.
* `convert_output` uses the formula `0xf - (2 * val)` which will give values in the range 15..-15.
* There are four channels summed together, giving a range of 60..-60.
* This is multiplied by the sound volume control values of 1..8, giving a range of 480..-480.
* You’re scaling against `32768 / 64`, i.e. 512, so you won’t get the full output range.  Wouldn’t it be better to use `outputl.put_int(sampindex, left, 15 * 4 * (1 + 7));` or something so the maximum output values will map to the full output range?